File: CatchOPP/CommunityMicroService/src/main/java/tn/esprit/communitymicroservice/services/ReportService.java
package tn.esprit.communitymicroservice.services;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import tn.esprit.communitymicroservice.entities.*;
import tn.esprit.communitymicroservice.repositories.CommentRepository;
import tn.esprit.communitymicroservice.repositories.PostRepository;
import tn.esprit.communitymicroservice.repositories.ReportRepository;

import java.util.List;

@Service
public class ReportService {

    @Autowired
    private ReportRepository reportRepository;

    @Autowired
    private PostRepository postRepository;

    @Autowired
    private CommentRepository commentRepository;

    // ── Create a report ───────────────────────────────────────────────────
    public Report createReport(Long reporterId, ReportTargetType targetType,
                               Long targetId, String reason) {

        // Validate target exists
        if (targetType == ReportTargetType.POST) {
            postRepository.findById(targetId)
                    .orElseThrow(() -> new RuntimeException("Post not found: " + targetId));
        } else if (targetType == ReportTargetType.COMMENT) {
            commentRepository.findById(targetId)
                    .orElseThrow(() -> new RuntimeException("Comment not found: " + targetId));
        }

        // Prevent duplicate reports from the same user on the same target
        if (reportRepository.existsByReporterIdAndTargetTypeAndTargetId(
                reporterId, targetType, targetId)) {
            throw new RuntimeException("You have already reported this content.");
        }

        Report report = new Report();
        report.setReporterId(reporterId);
        report.setTargetType(targetType);
        report.setTargetId(targetId);
        report.setReason(reason);
        report.setStatus(ReportStatus.PENDING);
        return reportRepository.save(report);
    }

    // ── Get all reports ───────────────────────────────────────────────────
    public List<Report> getAllReports() {
        return reportRepository.findAll();
    }

    // ── Get pending reports ───────────────────────────────────────────────
    public List<Report> getPendingReports() {
        return reportRepository.findByStatus(ReportStatus.PENDING);
    }

    // ── Get reports by target ─────────────────────────────────────────────
    public List<Report> getReportsByTarget(ReportTargetType targetType, Long targetId) {
        return reportRepository.findByTargetTypeAndTargetId(targetType, targetId);
    }

    // ── Get reports by reporter ───────────────────────────────────────────
    public List<Report> getReportsByReporter(Long reporterId) {
        return reportRepository.findByReporterId(reporterId);
    }

    // ── Dismiss a report (admin) ──────────────────────────────────────────
    public Report dismissReport(Long reportId) {
        Report report = reportRepository.findById(reportId)
                .orElseThrow(() -> new RuntimeException("Report not found: " + reportId));
        report.setStatus(ReportStatus.DISMISSED);
        return reportRepository.save(report);
    }

    // ── Delete reported content + mark report as REVIEWED (admin) ─────────
    public void deleteReportedContent(Long reportId) {
        Report report = reportRepository.findById(reportId)
                .orElseThrow(() -> new RuntimeException("Report not found: " + reportId));

        // Delete the actual content
        if (report.getTargetType() == ReportTargetType.POST) {
            postRepository.deleteById(report.getTargetId());
        } else if (report.getTargetType() == ReportTargetType.COMMENT) {
            commentRepository.deleteById(report.getTargetId());
        }

        // Mark all reports on this target as REVIEWED
        List<Report> relatedReports = reportRepository.findByTargetTypeAndTargetId(
                report.getTargetType(), report.getTargetId());
        relatedReports.forEach(r -> r.setStatus(ReportStatus.REVIEWED));
        reportRepository.saveAll(relatedReports);
    }

    // ── Get report by ID ──────────────────────────────────────────────────
    public Report getById(Long id) {
        return reportRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Report not found: " + id));
    }
}

File: FrontFreelanceApp/src/app/interfaces_events/report.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_CONFIG } from './api.config';

export interface CommunityReport {
  id: number;
  reporterId: number;
  targetType: 'POST' | 'COMMENT';
  targetId: number;
  reason: string;
  status: 'PENDING' | 'REVIEWED' | 'DISMISSED';
  createdAt: string;
}

@Injectable({ providedIn: 'root' })
export class ReportService {
  private apiUrl = `${API_CONFIG.BASE_URL}/api/reports`;

  constructor(private http: HttpClient) {}

  createReport(reporterId: number, targetType: 'POST' | 'COMMENT', targetId: number, reason: string): Observable<CommunityReport> {
    return this.http.post<CommunityReport>(this.apiUrl, { reporterId, targetType, targetId, reason });
  }

  getPendingReports(): Observable<CommunityReport[]> {
    return this.http.get<CommunityReport[]>(`${this.apiUrl}/pending`);
  }

  getAllReports(): Observable<CommunityReport[]> {
    return this.http.get<CommunityReport[]>(this.apiUrl);
  }

  deleteReportedContent(reportId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${reportId}/delete-content`);
  }

  dismissReport(reportId: number): Observable<CommunityReport> {
    return this.http.put<CommunityReport>(`${this.apiUrl}/${reportId}/dismiss`, {});
  }
}
